lib/irc/formatting.js

```diff
@@ -164,103 +214,67 @@ module.exports.htmlToIrc = function(html) {
-    var groups;
-    if (colorRegex.test(text)) {
-        groups = colorRegex.exec(text);
-        while (groups) {
-            // ignore bg for now (groups[2])
-
-            // this text includes the reset code so the code can be applied for
-            // each formatting code. We'll strip it later on.
-            var coloredText = groups[3];
-            var fontColor = colorCodesToHtmlNames[groups[1]];
-            if (fontColor) {
-                text = text.replace(
-                    groups[0],
-                    function() { // eslint-disable-line no-loop-func
-                        return '<font color="' + fontColor + '">' +
-                        coloredText +
-                        '</font>'
-                    }
-                );
-            }
-            else {
-                // unknown font colour
-                text = text.replace(groups[0], coloredText);
-            }
-            groups = colorRegex.exec(text);
-        }
-    }
+    // Escape HTML characters and add reset character to close all tags at the end.
+    text = escapeHtmlChars(text) + '\x0f';
 
-    // force there to always be a RESET code at the end of the text to make sure
-    // we close up all open style tags at the end.
-    if (text.length > 0 && text[text.length - 1] !== RESET_CODE) {
-        text = text + RESET_CODE;
-    }
+    // Replace all mIRC formatting characters.
+    // The color character can have arguments.
+    const colorRegex = /[\x02\x1d\x1f\x0f\x16]|\x03(\d{0,2})(?:,(\d{1,2}))?/g;
 
-    // Convert styles to bold/italics/underline
-    //
     // Maintain a small state machine of which tags are open so we can close the right
     // ones on RESET codes and toggle appropriately if they do the same code again.
-    let boldOpen = false;
-    let italicsOpen = false;
-    let underlineOpen = false;
-    for (let i = 0; i < text.length; i++) {
-        let char = text[i];
-        let charsToInsert = "";
-        switch (char) {
+    let state = Object.assign({}, STYLE_DEFAULT_STATE);
+
+    // Return message with codes replaced
+    return text.replace(colorRegex, function(match, fg, bg) {
+        let tags = '';
+
+        // Modify state with the current match
+        switch (match) {
             case STYLE_BOLD:
-                boldOpen = !boldOpen;
-                charsToInsert = boldOpen ? "<b>" : "</b>";
-                break;
+                return htmlTag(state, 'b');
+
             case STYLE_ITALICS:
-                italicsOpen = !italicsOpen;
-                charsToInsert = italicsOpen ? "<i>" : "</i>";
-                break;
+                return htmlTag(state, 'i');
+
             case STYLE_UNDERLINE:
-                underlineOpen = !underlineOpen;
-                charsToInsert = underlineOpen ? "<u>" : "</u>";
-                break;
+                return htmlTag(state, 'u');
+
+            case REVERSE_CODE:
+                state.color = [state.bcolor, state.bcolor = state.color][0];
```

Comment:
Style: What are you actually doing on this line? It looks like you're doing destructuring assignment coupled with array accesses. Can you please split this up onto multiple lines to aid readability.

Reply:
Swapping colours. This was a quick replacement for `[state.color, state.bcolor] = [state.bcolor, state.color]` (which is ES6 syntax). I've replaced it with something readable, because this isn't.